### PR TITLE
ECC: better protection when using encrypted memory

### PR DESCRIPTION
**Pull request by SparkiDev**
Added new ECC scalar multiplication implementation.

### CODE REVIEW

**Review by rizlik — APPROVED** [review]

File: wolfcrypt/src/ecc.c [hunk 1]
```diff
@@ -2925,6 +2931,178 @@ static int ecc_mulmod(const mp_int* k, ecc_point* P, ecc_point* Q,
...
+static int ecc_mulmod(const mp_int* k, ecc_point* P, ecc_point* Q,
+    ecc_point** R, mp_int* a, mp_int* modulus, mp_digit mp, WC_RNG* rng)
+{
+    int          err = MP_OKAY;
+    int          bytes = (mp_count_bits(modulus) + 7) / 8;
+    int          i;
+    int          j = 1;
+    int          cnt;
+    int          t = 0;
+    mp_int*      kt = R[4]->x;
```
Comment (dgarske) [inline wolfcrypt/src/ecc.c:2992]: Maybe use ‘ M_POINTS - 1’ instead of 4. Should we check for a NULL R arg?

File: wolfcrypt/src/ecc.c [hunk 2]
```diff
@@ -2925,6 +2931,178 @@ static int ecc_mulmod(const mp_int* k, ecc_point* P, ecc_point* Q,
...
+    /* R[1] = P */
+    if (err == MP_OKAY)
+        err = mp_copy(P->x, R[1]->x);
+    if (err == MP_OKAY)
+        err = mp_copy(P->y, R[1]->y);
+    if (err == MP_OKAY)
+        err = mp_copy(P->z, R[1]->z);
+
+    /* Randomize z ordinates to obfuscate timing. */
+    if ((err == MP_OKAY) && (rng != NULL))
```
Comment (dgarske) [inline wolfcrypt/src/ecc.c:3017]: What if ‘rng’ is NULL? How does it affect this function?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/ecc.c (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c:2992]: Changed

File: wolfcrypt/src/ecc.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/ecc.c:3017]: No obfuscation is performed as public key operation is being performed.

**Review by dgarske — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Pull request merged**